fe/spark-dpp/src/main/java/com/starrocks/load/loadv2/dpp/SparkDpp.java

```diff
@@ -610,48 +610,90 @@ public boolean validateData(Object srcValue, EtlJobConfig.EtlColumn etlColumn, C
+        srcColumnsWithColumnsFromPath.addAll(dataSrcColumns);
+        if (fileGroup.columnsFromPath != null) {
+            srcColumnsWithColumnsFromPath.addAll(fileGroup.columnsFromPath);
+        }
+        return createScrSchema(srcColumnsWithColumnsFromPath);
+    }
+
     /**
      * Note: parameter fileUrl cannot contain any wildcards to prevent output of wrong results.
      * For example, if fileUrl is specified by regular expression like hdfs://some/p1=*",
      * column value will be extracted as "*" in {@link DppUtils#parseColumnsFromPath},
      * then "*" will likely be cast to wrong output according to its field type such as NULL for INT
-     * in {@link this#convertSrcDataframeToDstDataframe}
+     * in {@link SparkDpp#convertSrcDataframeToDstDataframe}
      */
     private Dataset<Row> loadDataFromPath(SparkSession spark,
                                           EtlJobConfig.EtlFileGroup fileGroup,
                                           String fileUrl,
-                                          EtlJobConfig.EtlIndex baseIndex,
-                                          List<EtlJobConfig.EtlColumn> columns) throws SparkDppException {
+                                          EtlJobConfig.EtlIndex baseIndex) throws SparkDppException {
         List<String> columnValueFromPath = DppUtils.parseColumnsFromPath(fileUrl, fileGroup.columnsFromPath);
-        List<String> dataSrcColumns = fileGroup.fileFieldNames;
-        if (dataSrcColumns == null) {
-            // if there is no source columns info
-            // use base index columns as source columns
-            dataSrcColumns = new ArrayList<>();
-            for (EtlJobConfig.EtlColumn column : baseIndex.columns) {
-                dataSrcColumns.add(column.columnName);
+        StructType srcSchema = constructSrcSchema(fileGroup, baseIndex);
+        Dataset<Row> sourceData = null;
+        if (StringUtils.equalsIgnoreCase(fileGroup.fileFormat, "orc")) {
+            sourceData = spark.read().orc(fileUrl);
+        } else if (StringUtils.equalsIgnoreCase(fileGroup.fileFormat, "parquet")) {
+            sourceData = spark.read().parquet(fileUrl);
+        }
+        if (fileGroup.columnsFromPath != null) {
+            for (int i = 0; i < fileGroup.columnsFromPath.size(); i++) {
+                sourceData = sourceData.withColumn(fileGroup.columnsFromPath.get(i), functions.lit(columnValueFromPath.get(i)));
             }
         }
+        // check fields name and size
+        HashSet<String> schemaSet = new HashSet<>(Arrays.asList(sourceData.schema().fieldNames()));
+        schemaSet.addAll(Arrays.asList(srcSchema.fieldNames()));
+        if (sourceData.schema().size() != schemaSet.size()) {
+            throw new SparkDppException("The schema of file and load statement must be equal. " +
+                    "file schema: " + srcSchema.treeString() + ", table schema: " + sourceData.schema().treeString());
+        }
```

Comment:
Also need to check the field name

Reply:
The process of HashSet size comparison with source schema size already includes field name check.

Reply:
ok.
file schema is sourceData.schema() and table schema is srcSchema.

Reply:
fixed, PTAL